### PR TITLE
SONARJAVA-985 Fix forward reference of type parameters

### PR DESCRIPTION
**Pull request by benzonico**
Another error has been fix along the way when reading bounds of type parameter of types.

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/resolve/BytecodeVisitor.java [hunk 1]
```diff
@@ -318,11 +321,15 @@ public ReadGenericSignature() {
...
-      typeVariableSymbol = new JavaSymbol.TypeVariableJavaSymbol(name, classSymbol);
-      classSymbol.typeParameters.enter(typeVariableSymbol);
-      classSymbol.addTypeParameter((JavaType.TypeVariableJavaType) typeVariableSymbol.type);
-      bounds = Lists.newArrayList();
-      ((JavaType.TypeVariableJavaType) typeVariableSymbol.type).bounds = bounds;
+      List<JavaSymbol> lookup = classSymbol.typeParameters.lookup(name);
+      Preconditions.checkState(lookup.size() == 1, "found "+lookup.size());
+      typeVariableSymbol = (JavaSymbol.TypeVariableJavaSymbol) lookup.iterator().next();
+      bounds = ((JavaType.TypeVariableJavaType) typeVariableSymbol.type).bounds;
+//      typeVariableSymbol = new JavaSymbol.TypeVariableJavaSymbol(name, classSymbol);
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/BytecodeVisitor.java]: :new: MAJORThis block of commented-out lines of code should be removed.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: java-squid/src/main/java/org/sonar/java/resolve/BytecodeVisitor.java [hunk 2]
```diff
@@ -370,6 +399,39 @@ public void visitEnd() {
...
+      }else if (symbol.isMethodSymbol()) {
+        ((JavaSymbol.MethodJavaSymbol) symbol).typeParameters = new Scope(symbol);
+      }
+    }
+
+    @Override
+    public void visitFormalTypeParameter(String name) {
+      JavaSymbol.TypeVariableJavaSymbol typeVariableSymbol = new JavaSymbol.TypeVariableJavaSymbol(name, symbol);
+      if(symbol.isTypeSymbol()) {
+        JavaSymbol.TypeJavaSymbol classSymbol = (JavaSymbol.TypeJavaSymbol) symbol;
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/BytecodeVisitor.java]: :new: MAJORRename "classSymbol" which hides the field declared at line 43.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 new issues:
* 2 major

Details:
* [Rename "classSymbol" which hides the field declared at line 43.](https://github.com/SonarSource/sonar-java/blob/3a5f6577aa9cffc64759b39f4db8bdfb47c4be05/java-squid/src/main/java/org/sonar/java/resolve/BytecodeVisitor.java#L422)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AHiddenFieldCheck)
* [This block of commented-out lines of code should be removed.](https://github.com/SonarSource/sonar-java/blob/3a5f6577aa9cffc64759b39f4db8bdfb47c4be05/java-squid/src/main/java/org/sonar/java/resolve/BytecodeVisitor.java#L328)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

File: java-squid/src/main/java/org/sonar/java/resolve/BytecodeVisitor.java [hunk 3]
```diff
@@ -370,6 +394,39 @@ public void visitEnd() {
...
+    }
+
+    @Override
+    public void visitFormalTypeParameter(String name) {
+      JavaSymbol.TypeVariableJavaSymbol typeVariableSymbol = new JavaSymbol.TypeVariableJavaSymbol(name, symbol);
+      if(symbol.isTypeSymbol()) {
+        JavaSymbol.TypeJavaSymbol typeJavaSymbol = (JavaSymbol.TypeJavaSymbol) symbol;
+        typeJavaSymbol.typeParameters.enter(typeVariableSymbol);
+        typeJavaSymbol.addTypeParameter((JavaType.TypeVariableJavaType) typeVariableSymbol.type);
+        ((JavaType.TypeVariableJavaType) typeVariableSymbol.type).bounds = Lists.newArrayList();
```
Comment (m-g-sonar) [inline java-squid/src/main/java/org/sonar/java/resolve/BytecodeVisitor.java]: You can avoid the repetition of this line (lines 420 & 425) by setting the bounds before entering the `if`.

**Comment by m-g-sonar** [comment]
Except for previous comment, LGTM!

**Pull request closed without merging**